doc/developer/design/20210408_timelines.md

```diff
@@ -0,0 +1,76 @@
+# Timelines
+
+## Summary
+
+Some sources consist of data that doesn't have any timestamp information and we
+assign that data "real time" timestamps. Other kinds of sources, might have stronger
+opinions about the timestamps that data get bound to. In those cases, source
+implementations have a tradeoff. If they remap upstream timestamps to "real time"
+then the data from this source can be joined against other real time sources easily.
+However, this makes doing other operations, like temporal filters, on the source
+more difficult because now timestamps within the data (like a "delete-at-this-time" column)
+might not be meaningful when compared to the timestamp Materialized assigned the data.
+
+This design doc proposes to:
+
+  * Introduce the concept of `source timeline`s or `source time domain` (name TBD).
+    A `timeline` means roughly the same thing it does colloquially when people say
```

Comment:
This feels abstract to me. It's hard to evaluate this proposal without understanding how this maps to the current set of sources.

What sources would belong to the same timeline? All sources that are part of a single `CREATE SOURCES` statement? What about different `SOURCE`s from the same Kafka installation? How are `FILE` and `S3` sources handled? What about volatile sources?